### DISCUSSION TITLE
is there a way to easily import "langium-grammar.langium"; ?

### DISCUSSION THREAD

**Discussion by jdeantoni** [eclipse-langium/langium#899] [Q&A] [answered]
Hi,

I would like to propose a language that references ParserRules (or AbstractRules, to be defined later). 
I tried naively to do an `import "langium-grammar.langium";` in my grammar but this does not compile:
`/home/jdeanton/boulot/enseignements/L3IA_langageEtCompil/testLangium/sos/node_modules/langium/lib/utils/ast-util.js:85
    while (node.$container) {`
`                      ^`

`TypeError: Cannot read properties of undefined (reading '$container')`

I guess this is because I'm not telling where is the langium grammar exactly ? I'm not sure if there exist a registry or something similar from which I could import the grammar.

Additionally, once I will be able to do so, I would like in my language to import a specific `langium` grammar (e.g. `import "hello.langium"`. I used to use `importURI` in Xtext. Is there something similar ? Should I load myself the file in the scope provider and provision the ParserRules instances in the local scope manually ?  

Thanks in advance !

**Comment by msujew** [comment, marked as answer]
Hey @jdeantoni,

I'm a bit suprised by Langium spitting out an internal error for this (it should show a validation message instead). In theory you should be able to reference it by using something like `import "../node_modules/langium/src/language-server/langium-grammar";` (with as many `..` as needed for your directory structure). The language server might complain until you remove the `node_modules` from the ignore pattern of the langium vscode settings.

Note that compared to Xtext, this effectively only copies the grammar rules and the interfaces/types and does not create actual relationships between the original objects of the language. You can of course still parse `.langium` files using your project and reference these elements in your language, but there's no global storage where Langium elements are exported to so they're available for referencing. 

> Should I load myself the file in the scope provider and provision the ParserRules instances in the local scope manually ?

In theory it should be loaded together as part of your language server initialization and then scoped to the import. We do it similarly in our [own scope provider](https://github.com/langium/langium/blob/c7e15ee9663d37dab676dea4ffce479e3af0fea3/packages/langium/src/grammar/references/grammar-scope.ts#L54-L66).

**Comment by jdeantoni** [comment]
Thanks @msujew, that was very helpfull. Just a quick note, the link to be used is : `import "../../node_modules/langium/src/grammar/langium-grammar"`

With this import, I'm able to write down correctly my grammar. However, I was not able to correctly set up my scope provider by looking at yours. I'm able to retrieve the URI I want to load int the function but then some magic appeared in your scope provider since the elements seems to be already in the `indexManager` while mines are not there. Any idea why ? I did not understand how it was managed. 

Thanks !

**Reply by msujew** [reply]
How deep is your object structure for langium grammar elements? The default scope computation only puts elements into the index that are either the root note or a direct child of the root node. Everything else needs to be added using an override.

**Reply by jdeantoni** [reply]
It's actually flat for now. Here is the minimal example I try to setup:
```
test1
import "./simple-l.langium"

{
    var zaza on Person
}
```

where `simple-l.langium` is the hello world example grammar.

When I look what is inside the `indexManager`, I only found `test1` and `zaza`.
Since I have access in the scope provider to the `importedURI`, I guess I could ask explicitly for loading the document. If this is a way to go, could you please tell me what API is the most appropriate ?

**Reply by msujew** [reply]
Ah, I see. As mentioned in my [other answer](https://github.com/langium/langium/discussions/899#discussioncomment-4881248), Langium behaves quite differently compared to Xtext in regards to grammar imports. Since each language server is its own process, there's no reasonable way to share the index between languages. Langium will never expose its own index to other language servers. You actually need to create a multi-language project (see [here for an example](https://github.com/langium/langium/tree/main/examples/requirements)) and load the `.langium` files as part of your workspace. Afterwards, you should be able to implement a very similar scoping system to our own.

**Reply by jdeantoni** [reply]
oh I see, thanks it is slowly clarifying in my mind :).

 I'll carefully check the multi language example. Just a point before to dig up; the `langium` language actually already exists and corresponding artifacts are actually imported (i.e. in `node_modules`). 
 
Does this mean that I "only" need to clone the https://github.com/langium/langium/blob/main/examples/requirements/src/language-server/requirements-and-tests-lang-module.ts mechanism, where if I understand correctly it injects services from both languages ?

**Reply by msujew** [reply]
Yes, you should be able to reuse the `LangiumGrammarModule` in your multi-language module, so that you don't need to re-implement everything that's defined in there.

**Reply by jdeantoni** [reply]
Thanks it "almost" works correctly. Actually I have the appropriate feature but some errors popup as soon as I type:
```
[Error - 8:41:54 PM] Request textDocument/semanticTokens/range failed.
  Message: No semantic token provider registered
  Code: 0 
[Error - 8:41:55 PM] Request textDocument/semanticTokens/full failed.
  Message: No semantic token provider registered
  Code: 0 
```

any idea what I did wrong or where should I look at ?

**Reply by msujew** [reply]
Ah, yeah that's a bit annoying. Basically we send to the language client (vscode) that the language server supports semantic token (semantic syntax highlighting) requests. However, we already do that when one language of the language server supports that (the langium language does).

When the other language doesn't, it will yield this error, since vscode will still try to gather semantic token information although there is no provider registered for your language. You can easily register a null implementation though that simply extends the `SemanticTokenProvider` without an actual implementation.

We might actually want to change this, to just return an empty result in case the language doesn't have a provider for that. Thanks for bringing this up!

**Reply by jdeantoni** [reply]
Thanks a lot for your help ! It's working perfectly now !

**Reply by jdeantoni** [reply]
Sorry, not sure if I should open a new discussion or continue this one, but my new question is related to previous ones.
The editor is working correctly but I'm not able to run the generation. I have the following error before entering in the `generateAction `:
```
$> ../sos/bin/cli generate test1.sos 
There are validation errors:
line 5: Could not resolve reference to ParserRule named 'Person'. [Person]
```
I tried to do like for the requirements model multi language example but since I guess I missed something.

**Reply by msujew** [reply]
All good, it keeps the spirit of this discussions, since it's still about the import mechanism. The CLI works different in the sense that it does not automatically pick up a whole workspace; instead it only starts with a single entry file and then goes on to import more and more files as necessary (depending on their imports). We have a special loading mechanism in our CLI for that. See:

https://github.com/langium/langium/blob/929e261de07d4591e9faa4fd489875a991151738/packages/langium-cli/src/generate.ts#L63-L80

**Reply by jdeantoni** [reply]
Thanks, that's more or less what I tried to setup in the `cli-util.ts` but my code is not even called. Everything is as if it fails before entering code written on my side.

so, please, where should I inject the eager load mechanism ?

**Reply by msujew** [reply]
Assuming you are using the yeoman generator template, you code needs to run before the validation runs:

https://github.com/langium/langium/blob/929e261de07d4591e9faa4fd489875a991151738/packages/generator-langium/langium-template/src/cli/cli-util.ts#L22-L31

**Reply by jdeantoni** [reply]
It stops before that (or should I need to recompile the `cli` bin ?)
I already tried changing the error message to something else but not changes. I also tried console.logging  in the entry of the `generateAction` but without success

**Reply by msujew** [reply]
You will need to run the typescript compiler on the project after any changes. The `bin` folder only contains a simple `.js` file, but the `cli` folder needs to be compiled. I believe the generated project comes with a `watch` script can will automatically recompile any changes on disk.

**Reply by jdeantoni** [reply]
ok thanks, that's what I did...

Here is what I tried:
```
export const generateAction = async (fileName: string, opts: GenerateOptions): Promise<void> => {
    console.log("######## 1")
    throw new Error("SoS ERROR");
}
```
then 
```
const fileExtensions = StructuralOperationalSemanticsLanguageMetaData.fileExtensions.join(', ');
    program
        .command('generate')
        .argument('<file>', `source file (possible file extensions: ${fileExtensions})`)
        .option('-d, --destination <dir>', 'destination directory of generating')
        .description('generates JavaScript code that prints "Hello, {name}!" for each greeting in a source file')
        .action(generateAction);
```

and finally calling in the terminal like: `../sos/bin/cli generate test1.sos`
The js files in the `out` folder are compiled correctly. 

I should have miss a silly point (I'm not a natural TS developper)

**Reply by msujew** [reply]
Um, absolutely no idea what could be wrong there. Do you have a public repo at hand to reproduce this?

**Reply by jdeantoni** [reply]
yes, sure, thanks. Here it is: 
[https://github.com/jdeantoni/SoS.lang](https://github.com/jdeantoni/SoS.lang)

**Reply by msujew** [reply]
I'm not sure what's wrong there, it works for me right out of the box ¯\\\_(ツ)\_/¯

![image](https://user-images.githubusercontent.com/4377073/217532280-8fb715cc-3366-454f-b292-706a86b17d24.png)

**Reply by jdeantoni** [reply]
OK thanks so I guess this is some mess on my side. I'll check that ! thanks !

**Reply by jdeantoni** [reply]
oh ! sorry !! it was a mess in my folders and workspaces ! Sorry for that

**Reply by msujew** [reply]
All good. I'm glad things have worked out :)
